internal/controller/storageclient_controller.go

```diff
@@ -782,23 +782,31 @@ func removeStorageClaimAsOwner(obj client.Object) {
 func (r *storageClientReconcile) EnsureClusterScopedResource(obj client.Object, extRes *providerpb.ExternalResource, useReplace bool) error {
 	obj.SetName(extRes.Name)
 
-	if err := json.Unmarshal(extRes.Data, obj); err != nil {
-		return fmt.Errorf("failed to unmarshal %s configuration response: %v", obj.GetName(), err)
-	}
-	removeStorageClaimAsOwner(obj)
-	if err := r.own(obj); err != nil {
-		return fmt.Errorf("failed to own %s resource: %v", obj.GetName(), err)
-	}
-	utils.AddLabels(obj, extRes.Labels)
-	utils.AddAnnotations(obj, extRes.Annotations)
-
 	var err error
 	if useReplace {
 		err = utils.CreateOrReplace(r.ctx, r.Client, obj, func() error {
+			if err := json.Unmarshal(extRes.Data, obj); err != nil {
+				return fmt.Errorf("failed to unmarshal %s configuration response: %v", obj.GetName(), err)
+			}
+			removeStorageClaimAsOwner(obj)
+			if err := r.own(obj); err != nil {
+				return fmt.Errorf("failed to own %s resource: %v", obj.GetName(), err)
+			}
+			utils.AddLabels(obj, extRes.Labels)
+			utils.AddAnnotations(obj, extRes.Annotations)
 			return nil
 		})
 	} else {
 		err = r.createOrUpdate(obj, func() error {
+			if err := json.Unmarshal(extRes.Data, obj); err != nil {
+				return fmt.Errorf("failed to unmarshal %s configuration response: %v", obj.GetName(), err)
+			}
+			removeStorageClaimAsOwner(obj)
+			if err := r.own(obj); err != nil {
+				return fmt.Errorf("failed to own %s resource: %v", obj.GetName(), err)
+			}
+			utils.AddLabels(obj, extRes.Labels)
+			utils.AddAnnotations(obj, extRes.Annotations)
```

Comment:
This change creates duplicate code that need to be maintained in multiple places. As an alternative I suggest the creation of the mutation function at the same place where the original code was then passing it to `createOrUpdate` or `createOrReplace`

```suggestion
	mutateFunc := func() error { 
		if err := json.Unmarshal(extRes.Data, obj); err != nil {
			return fmt.Errorf("failed to unmarshal %s configuration response: %v", obj.GetName(), err)
		}
		removeStorageClaimAsOwner(obj)
		if err := r.own(obj); err != nil {
			return fmt.Errorf("failed to own %s resource: %v", obj.GetName(), err)
		}
		utils.AddLabels(obj, extRes.Labels)
		utils.AddAnnotations(obj, extRes.Annotations)
	}
	
	var err error
	if useReplace {
		err = utils.CreateOrReplace(r.ctx, r.Client, obj, mutateFunc)
	} else {
		err = r.createOrUpdate(obj, mutateFunc)
	}

```

Reply:
done